### PR TITLE
IGMP support

### PR DESCRIPTION
**Pull request by podhrmic**
related to #51 

I am seeking review early as recommended so I can easily fix and move things around.

This daily snapshot contains:

1. adds optional feature `protocol-igmp` so the multicast is optional
2. starts a minimal implementation of IGMPv2 protocol (v3 should be a future expansion)

As mentioned, this is not a full PR as things are not working yet, but I am looking for feedback/suggestions!

### CODE REVIEW

File: src/wire/igmp.rs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+use {Error, Result};
+use phy::ChecksumCapabilities;
+use super::ip::checksum;
+use super::{Ipv4Packet, Ipv4Repr};
+
+use wire::Ipv4Address;
+
+enum_with_unknown! {
+    /// Internet protocol control message type.
+    pub doc enum Igmpv2MessageType(u8) {
```
Comment (whitequark) [inline src/wire/igmp.rs]: ... and when the file is called `igmpv2` you can remove the obnoxious prefixes. See how the other protocols do it.

File: src/phy/mod.rs [hunk 2]
```diff
@@ -179,6 +179,7 @@ pub struct ChecksumCapabilities {
     pub udpv4: Checksum,
     pub tcpv4: Checksum,
     pub icmpv4: Checksum,
+    pub igmp: Checksum,
```
Comment (whitequark) [inline src/phy/mod.rs]: I don't think there's any NIC that does IGMP checksum offload. Even ICMP was a surprise to me.

File: src/wire/igmp.rs [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+            &Igmpv2MessageType::MembershipReportV2 => write!(f, "version 2 membership report "),
+            &Igmpv2MessageType::LeaveGroup => write!(f, "leave group"),
+            &Igmpv2MessageType::MembershipReportV1 => write!(f, "version 1 membership report"),
+        }
+    }
+}
+
+
+impl<T: AsRef<[u8]>> Packet<T> {
+    /// Imbue a raw octet buffer with ICMPv4 packet structure.
```
Comment (whitequark) [inline src/wire/igmp.rs]: Wrong doc comment.

File: src/wire/igmp.rs [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+    pub fn new_checked(buffer: T) -> Result<Packet<T>> {
+        let packet = Self::new(buffer);
+        packet.check_len()?;
+        Ok(packet)
+    }
+
+    /// Ensure that no accessor method will panic if called.
+    /// Returns `Err(Error::Truncated)` if the buffer is too short.
+    ///
+    /// The result of this check is invalidated by calling [set_header_len].
```
Comment (whitequark) [inline src/wire/igmp.rs]: The header length is fixed, so there's nothing to invalidate.

File: src/wire/igmp.rs [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+                Err(Error::Truncated)
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    /// Return the header length.
+    /// The result depends on the value of the message type field.
+    pub fn header_len(&self) -> usize {
```
Comment (whitequark) [inline src/wire/igmp.rs]: This is not necessary also.

File: src/wire/igmp.rs [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+    /// Return the message type field.
+    #[inline]
+    pub fn msg_type(&self) -> Igmpv2MessageType {
+        let data = self.buffer.as_ref();
+        Igmpv2MessageType::from(data[field::TYPE])
+    }
+
+    /// Return the Max reponse time
+    #[inline]
+    pub fn get_max_resp_time(&self) -> u8 {
```
Comment (whitequark) [inline src/wire/igmp.rs]: Without `get_`.

File: src/wire/igmp.rs [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+        let data = self.buffer.as_ref();
+        checksum::data(data) == !0
+    }
+}
+
+
+impl<'a, T: AsRef<[u8]> + ?Sized> Packet<&'a T> {
+    /// Return a pointer to the type-specific data.
+    #[inline]
+    pub fn data(&self) -> &'a [u8] {
```
Comment (whitequark) [inline src/wire/igmp.rs]: No data in IGMP packets.

File: src/iface/ethernet.rs [hunk 8]
```diff
@@ -345,6 +349,10 @@ impl<'a, 'b, 'c, DeviceT: Device + 'a> Interface<'a, 'b, 'c, DeviceT> {
             IpProtocol::Icmp =>
                 self.process_icmpv4(ipv4_repr, ip_payload),
 
+			#[cfg(feature = "protocol-igmp")]
```
[the remark this answers is not in the document]
Comment (podhrmic) [reply src/iface/ethernet.rs]: Fixed

File: src/phy/mod.rs (hunk 2 above)
Comment (podhrmic) [reply src/phy/mod.rs]: Fixed

File: src/wire/igmp.rs (hunk 7 above)
Comment (podhrmic) [reply src/wire/igmp.rs]: Fixed

File: src/wire/igmp.rs (hunk 6 above)
Comment (podhrmic) [reply src/wire/igmp.rs]: Fixed

File: src/wire/igmp.rs (hunk 5 above)
Comment (podhrmic) [reply src/wire/igmp.rs]: Fixed

File: src/wire/igmp.rs (hunk 3 above)
Comment (podhrmic) [reply src/wire/igmp.rs]: Fixed

File: src/wire/igmp.rs (hunk 1 above)
Comment (podhrmic) [reply src/wire/igmp.rs]: Fixed to look like for ICMP packets

File: src/wire/igmp.rs [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+
+use {Error, Result};
+use phy::ChecksumCapabilities;
+use super::ip::checksum;
+use super::{Ipv4Packet, Ipv4Repr};
+
+use wire::Ipv4Address;
+
+enum_with_unknown! {
+    /// Internet protocol control message type.
```
[the remark this answers is not in the document]
Comment (podhrmic) [reply src/wire/igmp.rs]: Fixed

File: src/wire/igmp.rs [hunk 10]
```diff
@@ -0,0 +1,262 @@
+/// Internet Group Management Protocol v2
```
[the remark this answers is not in the document]
Comment (podhrmic) [reply src/wire/igmp.rs]: Fixed

**Comment by podhrmic** [comment]
OK, adding another daily snapshot. Added parse/emit functions and simple tests to verify that parsing/creating IGMP packets works.

The next step is to:

1. test sending/receiving IGMP packets over Ethernet iface
2. add application logic for subscribing/leaving groups

@whitequark let me know if you prefer daily snapshots or rather larger chunks of code.

**Comment by podhrmic** [comment]
Closing for now, will reopen with more complete feature set.

**Pull request closed without merging**